File: server/src/commonMain/kotlin/com/supermetroid/parser/GameStateParser.kt
package com.supermetroid.parser

import com.supermetroid.model.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Super Metroid Game State Parser
 * Converts raw memory data into structured game state
 * Equivalent to Python's SuperMetroidGameStateParser
 */
class GameStateParser {
    private val mutex = Mutex()
    
    // Persistent state for Mother Brain phases - once detected, stays detected
    private var motherBrainPhaseState = mutableMapOf(
        "mb1_detected" to false,
        "mb2_detected" to false
    )
    
    // Super Metroid memory layout
    private val memoryMap = mapOf(
        "health" to 0x7E09C2,
        "max_health" to 0x7E09C4,
        "missiles" to 0x7E09C6,
        "max_missiles" to 0x7E09C8,
        "supers" to 0x7E09CA,
        "max_supers" to 0x7E09CC,
        "power_bombs" to 0x7E09CE,
        "max_power_bombs" to 0x7E09D0,
        "reserve_energy" to 0x7E09D6,
        "max_reserve_energy" to 0x7E09D4,
        "room_id" to 0x7E079B,
        "area_id" to 0x7E079F,
        "game_state" to 0x7E0998,
        "player_x" to 0x7E0AF6,
        "player_y" to 0x7E0AFA,
        "items" to 0x7E09A4,
        "beams" to 0x7E09A8,
        "bosses" to 0x7ED828,
    )
    
    // Area names mapping
    private val areas = mapOf(
        0 to "Crateria",
        1 to "Brinstar", 
        2 to "Norfair",
        3 to "Wrecked Ship",
        4 to "Maridia",
        5 to "Tourian"
    )
    
    suspend fun parseCompleteGameState(memoryData: Map<String, ByteArray>): GameState = mutex.withLock {
        try {
            // Parse basic stats from bulk read
            val basicStats = memoryData["basic_stats"]
            val stats = if (basicStats != null && basicStats.size >= 22) {
                parseBasicStats(basicStats)
            } else {
                mapOf<String, Int>()
            }
            
            // Parse location data
            val locationData = parseLocationData(
                memoryData["room_id"],
                memoryData["area_id"],
                memoryData["game_state"],
                memoryData["player_x"],
                memoryData["player_y"]
            )
            
            // Parse items with reset detection
            val items = parseItems(
                memoryData["items"],
                locationData,
                stats["health"] ?: 0,
                stats["missiles"] ?: 0,
                stats["max_missiles"] ?: 0
            )
            
            // Parse beams with reset detection  
            val beams = parseBeams(
                memoryData["beams"],
                locationData,
                stats["health"] ?: 0,
                stats["missiles"] ?: 0,
                stats["max_missiles"] ?: 0
            )
            
            // Parse bosses
            val bosses = parseBosses(
                memoryData["main_bosses"],
                memoryData["crocomire"],
                memoryData["boss_plus_1"],
                memoryData["boss_plus_2"],
                memoryData["boss_plus_3"],
                memoryData["boss_plus_4"],
                memoryData["boss_plus_5"],
                memoryData["escape_timer_1"],
                memoryData["escape_timer_2"],
                memoryData["escape_timer_3"],
                memoryData["escape_timer_4"],
                locationData
            )
            
            return GameState(
                health = stats["health"] ?: 0,
                maxHealth = stats["max_health"] ?: 99,
                missiles = stats["missiles"] ?: 0,
                maxMissiles = stats["max_missiles"] ?: 0,
                supers = stats["supers"] ?: 0,
                maxSupers = stats["max_supers"] ?: 0,
                powerBombs = stats["power_bombs"] ?: 0,
                maxPowerBombs = stats["max_power_bombs"] ?: 0,
                reserveEnergy = stats["reserve_energy"] ?: 0,
                maxReserveEnergy = stats["max_reserve_energy"] ?: 0,
                roomId = locationData["room_id"] as? Int ?: 0,
                areaId = locationData["area_id"] as? Int ?: 0,
                areaName = locationData["area_name"] as? String ?: "Unknown",
                gameState = locationData["game_state"] as? Int ?: 0,
                playerX = locationData["player_x"] as? Int ?: 0,
                playerY = locationData["player_y"] as? Int ?: 0,
                items = items,
                beams = beams,
                bosses = bosses
            )
            
        } catch (e: Exception) {
            println("Error parsing game state: ${e.message}")
            GameState()
        }
    }
    
    private fun parseBasicStats(data: ByteArray): Map<String, Int> {
        if (data.size < 22) return emptyMap()
        
        return try {
            mapOf(
                "health" to data.readInt16LE(0),
                "max_health" to data.readInt16LE(2),
                "missiles" to data.readInt16LE(4),
                "max_missiles" to data.readInt16LE(6),
                "supers" to data.readInt16LE(8),
                "max_supers" to data.readInt16LE(10),
                "power_bombs" to data.readInt16LE(12),
                "max_power_bombs" to data.readInt16LE(14),
                "reserve_energy" to data.readInt16LE(18),
                "max_reserve_energy" to data.readInt16LE(16)
            )
        } catch (e: Exception) {
            println("Error parsing basic stats: ${e.message}")
            emptyMap()
        }
    }
    
    private fun parseLocationData(
        roomIdData: ByteArray?,
        areaIdData: ByteArray?,
        gameStateData: ByteArray?,
        playerXData: ByteArray?,
        playerYData: ByteArray?
    ): Map<String, Any> {
        return try {
            val roomId = roomIdData?.readInt16LE(0) ?: 0
            val areaId = areaIdData?.let { if (it.isNotEmpty()) it[0].toInt() and 0xFF else 0 } ?: 0
            val gameState = gameStateData?.readInt16LE(0) ?: 0
            val playerX = playerXData?.readInt16LE(0) ?: 0
            val playerY = playerYData?.readInt16LE(0) ?: 0
            
            mapOf(
                "room_id" to roomId,
                "area_id" to areaId,
                "area_name" to (areas[areaId] ?: "Unknown"),
                "game_state" to gameState,
                "player_x" to playerX,
                "player_y" to playerY
            )
        } catch (e: Exception) {
            println("Error parsing location data: ${e.message}")
            emptyMap()
        }
    }
    
    private fun shouldResetItemState(
        locationData: Map<String, Any>,
        health: Int,
        missiles: Int,
        maxMissiles: Int
    ): Boolean {
        val areaId = locationData["area_id"] as? Int ?: 0
        val roomId = locationData["room_id"] as? Int ?: 0
        
        // Reset scenarios - CONSERVATIVE:
        
        // 1. Intro scene (very specific early game indicators)
        val inStartingArea = areaId == 0 // Crateria
        val hasStartingHealth = health <= 99
        val inStartingRooms = roomId < 1000
        if (inStartingArea && hasStartingHealth && inStartingRooms) {
            println("🔄 ITEM STATE RESET: intro scene detected - Area:$areaId, Room:$roomId, Health:$health")
            return true
        }
        
        // 2. Very specific new game indicators only
        val definiteNewGame = (health == 99 && missiles == 0 && maxMissiles == 0 && roomId < 1000)
        if (definiteNewGame) {
            println("🔄 ITEM STATE RESET: new game detected - Area:$areaId, Room:$roomId, Health:$health, Missiles:$missiles/$maxMissiles")
            return true
        }
        
        return false
    }
    
    private fun parseItems(
        itemsData: ByteArray?,
        locationData: Map<String, Any>,
        health: Int,
        missiles: Int,
        maxMissiles: Int
    ): Map<String, Boolean> {
        if (itemsData == null || itemsData.size < 2) return emptyMap()
        
        // Check if we should reset item state
        if (shouldResetItemState(locationData, health, missiles, maxMissiles)) {
            return mapOf(
                "morph" to false,
                "bombs" to false,
                "varia" to false,
                "gravity" to false,
                "hijump" to false,
                "speed" to false,
                "spacejump" to false,
                "screw" to false,
                "spring" to false,
                "grapple" to false,
                "xray" to false
            )
        }
        
        val itemValue = itemsData.readInt16LE(0)
        
        return mapOf(
            "morph" to ((itemValue and 0x0004) != 0),
            "bombs" to ((itemValue and 0x1000) != 0),
            "varia" to ((itemValue and 0x0001) != 0),
            "gravity" to ((itemValue and 0x0020) != 0),
            "hijump" to ((itemValue and 0x0100) != 0),
            "speed" to ((itemValue and 0x2000) != 0),
            "spacejump" to ((itemValue and 0x0200) != 0),
            "screw" to ((itemValue and 0x0008) != 0),
            "spring" to ((itemValue and 0x0002) != 0),
            "grapple" to ((itemValue and 0x4000) != 0),
            "xray" to ((itemValue and 0x8000) != 0)
        )
    }
    
    private fun parseBeams(
        beamsData: ByteArray?,
        locationData: Map<String, Any>,
        health: Int,
        missiles: Int,
        maxMissiles: Int
    ): Map<String, Boolean> {
        if (beamsData == null || beamsData.size < 2) return emptyMap()
        
        // Check if we should reset beam state (same logic as items)
        if (shouldResetItemState(locationData, health, missiles, maxMissiles)) {
            println("🔄 BEAM STATE RESET: Resetting to starting state (no beams)")
            return mapOf(
                "charge" to false,  // Charge beam must be collected
                "ice" to false,
                "wave" to false,
                "spazer" to false,
                "plasma" to false,
                "hyper" to false
            )
        }
        
        val beamValue = beamsData.readInt16LE(0)
        
        return mapOf(
            "charge" to ((beamValue and 0x1000) != 0),
            "ice" to ((beamValue and 0x0002) != 0),
            "wave" to ((beamValue and 0x0001) != 0),
            "spazer" to ((beamValue and 0x0004) != 0),
            "plasma" to ((beamValue and 0x0008) != 0),
            "hyper" to ((beamValue and 0x0010) != 0)
        )
    }
    
    private fun parseBosses(
        mainBossesData: ByteArray?,
        crocomireData: ByteArray?,
        bossPlus1Data: ByteArray?,
        bossPlus2Data: ByteArray?,
        bossPlus3Data: ByteArray?,
        bossPlus4Data: ByteArray?,
        bossPlus5Data: ByteArray?,
        escapeTimer1Data: ByteArray?,
        escapeTimer2Data: ByteArray?,
        escapeTimer3Data: ByteArray?,
        escapeTimer4Data: ByteArray?,
        locationData: Map<String, Any>
    ): Map<String, Boolean> {
        // Simplified boss parsing - full implementation would match Python logic
        val mainBosses = mainBossesData?.readInt16LE(0) ?: 0
        
        return mapOf(
            "bomb_torizo" to ((mainBosses and 0x0004) != 0),
            "kraid" to ((mainBosses and 0x0001) != 0),
            "spore_spawn" to ((mainBosses and 0x0002) != 0),
            "crocomire" to ((mainBosses and 0x0008) != 0),
            "botwoon" to ((mainBosses and 0x0010) != 0),
            "phantoon" to ((mainBosses and 0x0100) != 0),
            "draygon" to ((mainBosses and 0x0200) != 0),
            "ridley" to ((mainBosses and 0x0400) != 0),
            "golden_torizo" to ((mainBosses and 0x0800) != 0),
            "mother_brain_1" to motherBrainPhaseState["mb1_detected"]!!,
            "mother_brain_2" to motherBrainPhaseState["mb2_detected"]!!
        )
    }
    
    suspend fun resetMotherBrainCache() = mutex.withLock {
        motherBrainPhaseState["mb1_detected"] = false
        motherBrainPhaseState["mb2_detected"] = false
        println("🔄 MB cache reset to default (not detected)")
    }
    
    // Extension function to read little-endian 16-bit integers from ByteArray
    private fun ByteArray.readInt16LE(offset: Int): Int {
        if (offset + 1 >= size) return 0
        return (this[offset].toInt() and 0xFF) or ((this[offset + 1].toInt() and 0xFF) shl 8)
    }
} 

File: src/client/gameStateParser.ts
import type { GameStats } from '../types/gameState';

/**
 * Frontend Game State Parser
 * Converts raw memory data into structured game state for the browser
 */
export class FrontendGameStateParser {
  // Persistent state for Mother Brain phases - once detected, stays detected
  private motherBrainPhaseState = {
    mb1_detected: false,
    mb2_detected: false
  };

  // Area names mapping
  private readonly areas = {
    0: 'Crateria',
    1: 'Brinstar',
    2: 'Norfair',
    3: 'Wrecked Ship',
    4: 'Maridia',
    5: 'Tourian'
  };

  /**
   * Parse complete game state from memory data
   */
  parseCompleteGameState(memoryData: Record<string, Uint8Array | null>): GameStats {
    try {
      // Parse basic stats from bulk read
      const basicStats = memoryData['basic_stats'];
      const stats = basicStats && basicStats.length >= 22 
        ? this.parseBasicStats(basicStats) 
        : {};

      // Parse location data
      const locationData = this.parseLocationData(
        memoryData['room_id'],
        memoryData['area_id'],
        memoryData['game_state'],
        memoryData['player_x'],
        memoryData['player_y']
      );

      // Parse items with reset detection
      const items = this.parseItems(
        memoryData['items'],
        locationData,
        stats.health || 0,
        stats.missiles || 0,
        stats.max_missiles || 0
      );

      // Parse beams with reset detection
      const beams = this.parseBeams(
        memoryData['beams'],
        locationData,
        stats.health || 0,
        stats.missiles || 0,
        stats.max_missiles || 0
      );

      // Parse bosses
      const bosses = this.parseBosses(
        memoryData['main_bosses'],
        memoryData['crocomire'],
        memoryData['boss_plus_1'],
        memoryData['boss_plus_2'],
        memoryData['boss_plus_3'],
        memoryData['boss_plus_4'],
        memoryData['boss_plus_5'],
        memoryData['escape_timer_1'],
        memoryData['escape_timer_2'],
        memoryData['escape_timer_3'],
        memoryData['escape_timer_4'],
        locationData,
        memoryData['ship_ai'],
        memoryData['event_flags']
      );

      return {
        health: stats.health || 0,
        max_health: stats.max_health || 99,
        missiles: stats.missiles || 0,
        max_missiles: stats.max_missiles || 0,
        supers: stats.supers || 0,
        max_supers: stats.max_supers || 0,
        power_bombs: stats.power_bombs || 0,
        max_power_bombs: stats.max_power_bombs || 0,
        reserve_energy: stats.reserve_energy || 0,
        max_reserve_energy: stats.max_reserve_energy || 0,
        room_id: locationData.room_id || 0,
        area_id: locationData.area_id || 0,
        area_name: locationData.area_name || 'Unknown',
        game_state: locationData.game_state || 0,
        player_x: locationData.player_x || 0,
        player_y: locationData.player_y || 0,
        items: {
          morph: items.morph || false,
          bombs: items.bombs || false,
          varia: items.varia || false,
          gravity: items.gravity || false,
          hijump: items.hijump || false,
          speed: items.speed || false,
          space: items.space || false,
          screw: items.screw || false,
          grapple: items.grapple || false,
          xray: items.xray || false,
          spring: items.spring || false,
        },
        beams: {
          charge: beams.charge || false,
          ice: beams.ice || false,
          wave: beams.wave || false,
          spazer: beams.spazer || false,
          plasma: beams.plasma || false,
          hyper: beams.hyper || false,
        },
        bosses: {
          bomb_torizo: bosses.bomb_torizo || false,
          kraid: bosses.kraid || false,
          spore_spawn: bosses.spore_spawn || false,
          mother_brain: bosses.mother_brain || false,
          crocomire: bosses.crocomire || false,
          phantoon: bosses.phantoon || false,
          botwoon: bosses.botwoon || false,
          draygon: bosses.draygon || false,
          ridley: bosses.ridley || false,
          golden_torizo: bosses.golden_torizo || false,
          mother_brain_1: bosses.mother_brain_1 || false,
          mother_brain_2: bosses.mother_brain_2 || false,
          samus_ship: bosses.samus_ship || false,
        }
      };
    } catch (err) {
      console.log(`Error parsing game state: ${err}`);
      return this.getEmptyGameState();
    }
  }

  /**
   * Parse basic stats from memory data
   */
  private parseBasicStats(data: Uint8Array): Record<string, number> {
    if (data.length < 22) return {};

    try {
      return {
        health: this.readInt16LE(data, 0),
        max_health: this.readInt16LE(data, 2),
        missiles: this.readInt16LE(data, 4),
        max_missiles: this.readInt16LE(data, 6),
        supers: this.readInt16LE(data, 8),
        max_supers: this.readInt16LE(data, 10),
        power_bombs: this.readInt16LE(data, 12),
        max_power_bombs: this.readInt16LE(data, 14),
        max_reserve_energy: this.readInt16LE(data, 18),
        reserve_energy: this.readInt16LE(data, 20)
      };
    } catch (err) {
      console.log(`Error parsing basic stats: ${err}`);
      return {};
    }
  }

  /**
   * Parse location data
   */
  private parseLocationData(
    roomIdData: Uint8Array | null,
    areaIdData: Uint8Array | null,
    gameStateData: Uint8Array | null,
    playerXData: Uint8Array | null,
    playerYData: Uint8Array | null
  ): Record<string, any> {
    try {
      const roomId = roomIdData ? this.readInt16LE(roomIdData, 0) : 0;
      const areaId = areaIdData && areaIdData.length > 0 ? areaIdData[0] : 0;
      const gameState = gameStateData ? this.readInt16LE(gameStateData, 0) : 0;
      const playerX = playerXData ? this.readInt16LE(playerXData, 0) : 0;
      const playerY = playerYData ? this.readInt16LE(playerYData, 0) : 0;

      return {
        room_id: roomId,
        area_id: areaId,
        area_name: this.areas[areaId as keyof typeof this.areas] || 'Unknown',
        game_state: gameState,
        player_x: playerX,
        player_y: playerY
      };
    } catch (err) {
      console.log(`Error parsing location data: ${err}`);
      return {};
    }
  }

  /**
   * Check if we should reset item state (new game detection)
   */
  private shouldResetItemState(
    locationData: Record<string, any>,
    health: number,
    missiles: number,
    maxMissiles: number
  ): boolean {
    const areaId = locationData.area_id || 0;
    const roomId = locationData.room_id || 0;

    // Reset scenarios - CONSERVATIVE:

    // 1. Intro scene (very specific early game indicators)
    const inStartingArea = areaId === 0; // Crateria
    const hasStartingHealth = health <= 99;
    const inStartingRooms = roomId < 1000;
    if (inStartingArea && hasStartingHealth && inStartingRooms) {
      console.log(`🔄 ITEM STATE RESET: intro scene detected - Area:${areaId}, Room:${roomId}, Health:${health}`);
      return true;
    }

    // 2. Very specific new game indicators only
    const definiteNewGame = (health === 99 && missiles === 0 && maxMissiles === 0 && roomId < 1000);
    if (definiteNewGame) {
      console.log(`🔄 ITEM STATE RESET: new game detected - Area:${areaId}, Room:${roomId}, Health:${health}, Missiles:${missiles}/${maxMissiles}`);
      return true;
    }

    return false;
  }

  /**
   * Parse items from memory data
   */
  private parseItems(
    itemsData: Uint8Array | null,
    locationData: Record<string, any>,
    health: number,
    missiles: number,
    maxMissiles: number
  ): Record<string, boolean> {
    if (!itemsData || itemsData.length < 2) return {};

    // Check if we should reset item state
    if (this.shouldResetItemState(locationData, health, missiles, maxMissiles)) {
      return {
        morph: false,
        bombs: false,
        varia: false,
        gravity: false,
        hijump: false,
        speed: false,
        space: false,
        screw: false,
        spring: false,
        grapple: false,
        xray: false
      };
    }

    const itemValue = this.readInt16LE(itemsData, 0);

    return {
      morph: (itemValue & 0x0004) !== 0,
      bombs: (itemValue & 0x1000) !== 0,
      varia: (itemValue & 0x0001) !== 0,
      gravity: (itemValue & 0x0020) !== 0,
      hijump: (itemValue & 0x0100) !== 0,
      speed: (itemValue & 0x2000) !== 0,
      space: (itemValue & 0x0200) !== 0,
      screw: (itemValue & 0x0008) !== 0,
      spring: (itemValue & 0x0002) !== 0,
      grapple: (itemValue & 0x4000) !== 0,
      xray: (itemValue & 0x8000) !== 0
    };
  }

  /**
   * Parse beams from memory data
   */
  private parseBeams(
    beamsData: Uint8Array | null,
    locationData: Record<string, any>,
    health: number,
    missiles: number,
    maxMissiles: number
  ): Record<string, boolean> {
    if (!beamsData || beamsData.length < 2) return {};

    // Check if we should reset beam state (same logic as items)
    if (this.shouldResetItemState(locationData, health, missiles, maxMissiles)) {
      console.log('🔄 BEAM STATE RESET: Resetting to starting state (no beams)');
      return {
        charge: false,  // Charge beam must be collected
        ice: false,
        wave: false,
        spazer: false,
        plasma: false,
        hyper: false
      };
    }

    const beamValue = this.readInt16LE(beamsData, 0);

    return {
      charge: (beamValue & 0x1000) !== 0,
      ice: (beamValue & 0x0002) !== 0,
      wave: (beamValue & 0x0001) !== 0,
      spazer: (beamValue & 0x0004) !== 0,
      plasma: (beamValue & 0x0008) !== 0,
      hyper: (beamValue & 0x0010) !== 0
    };
  }

  /**
   * Parse bosses from memory data (simplified version)
   */
  private parseBosses(
    mainBossesData: Uint8Array | null,
    crocomireData: Uint8Array | null,
    _bossPlus1Data: Uint8Array | null,
    _bossPlus2Data: Uint8Array | null,
    _bossPlus3Data: Uint8Array | null,
    _bossPlus4Data: Uint8Array | null,
    _bossPlus5Data: Uint8Array | null,
    _escapeTimer1Data: Uint8Array | null,
    _escapeTimer2Data: Uint8Array | null,
    _escapeTimer3Data: Uint8Array | null,
    _escapeTimer4Data: Uint8Array | null,
    _locationData: Record<string, any>,
    _shipAiData?: Uint8Array | null,
    _eventFlagsData?: Uint8Array | null
  ): Record<string, boolean> {
    // Basic boss flags - using simplified logic for now
    const mainBosses = mainBossesData ? this.readInt16LE(mainBossesData, 0) : 0;

    return {
      bomb_torizo: (mainBosses & 0x0004) !== 0,
      kraid: (mainBosses & 0x0001) !== 0,
      spore_spawn: (mainBosses & 0x0002) !== 0,
      mother_brain: (mainBosses & 0x0001) !== 0,
      crocomire: crocomireData ? (this.readInt16LE(crocomireData, 0) & 0x0001) !== 0 : false,
      phantoon: (mainBosses & 0x0100) !== 0,
      botwoon: (mainBosses & 0x0200) !== 0,
      draygon: (mainBosses & 0x0400) !== 0,
      ridley: (mainBosses & 0x0010) !== 0,
      golden_torizo: (mainBosses & 0x0020) !== 0,
      mother_brain_1: this.motherBrainPhaseState.mb1_detected,
      mother_brain_2: this.motherBrainPhaseState.mb2_detected,
      samus_ship: false // Placeholder for end game detection
    };
  }

  /**
   * Reset Mother Brain cache
   */
  resetMotherBrainCache(): void {
    this.motherBrainPhaseState.mb1_detected = false;
    this.motherBrainPhaseState.mb2_detected = false;
    console.log('🧠 Mother Brain cache reset to default (not detected)');
  }

  /**
   * Read 16-bit little-endian integer from Uint8Array
   */
  private readInt16LE(data: Uint8Array, offset: number): number {
    if (offset + 1 >= data.length) return 0;
    return data[offset] | (data[offset + 1] << 8);
  }

  /**
   * Get empty game state
   */
  private getEmptyGameState(): GameStats {
    return {
      health: 0,
      max_health: 99,
      missiles: 0,
      max_missiles: 0,
      supers: 0,
      max_supers: 0,
      power_bombs: 0,
      max_power_bombs: 0,
      reserve_energy: 0,
      max_reserve_energy: 0,
      room_id: 0,
      area_id: 0,
      area_name: 'Unknown',
      game_state: 0,
      player_x: 0,
      player_y: 0,
      items: {
        morph: false,
        bombs: false,
        varia: false,
        gravity: false,
        hijump: false,
        speed: false,
        space: false,
        screw: false,
        grapple: false,
        xray: false,
        spring: false,
      },
      beams: {
        charge: false,
        ice: false,
        wave: false,
        spazer: false,
        plasma: false,
        hyper: false,
      },
      bosses: {
        bomb_torizo: false,
        kraid: false,
        spore_spawn: false,
        mother_brain: false,
        crocomire: false,
        phantoon: false,
        botwoon: false,
        draygon: false,
        ridley: false,
        golden_torizo: false,
        mother_brain_1: false,
        mother_brain_2: false,
        samus_ship: false,
      }
    };
  }
}
